lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+
+    def _on_start(self, _: StartEvent) -> None:
+        # This will trigger the juju-systemd-notices daemon to
+        # emit a `service-slurmd-started` event.
+        systemd.service_start("slurmd")
+
+    def _on_stop(self, _: StopEvent) -> None:
+        # To stop the juju-systemd-notices service running in the background.
+        # .stop() also disables the juju-systemd-notices so that it does not
+        # start back up if the underlying machine is rebooted.
+        self._systemd_notices.stop()
+
+    def _on_slurmd_started(self, _: ServiceStartedEvent) -> None:
+        self.unit.status = ActiveStatus()
+        time.sleep(60)
+
+        # This will trigger the juju-systemd-notices daemon to
+        # emit a `service-slurmd-stopped` event.
+        systemd.service_stop("slurmd")
+
+    def _on_slurmd_stopped(self, _: ServiceStoppedEvent) -> None:
+        self.unit.status = BlockedStatus("slurmd not running")
+```
+"""
+
+__all__ = ["ServiceStartedEvent", "ServiceStoppedEvent", "SystemdNotices"]
+
+import argparse
+import asyncio
+import functools
+import logging
+import re
+import signal
+import subprocess
+import sys
+import textwrap
+from pathlib import Path
+from types import MappingProxyType
+from typing import List, Union
+
+from dbus_fast.aio import MessageBus
+from dbus_fast.constants import BusType, MessageType
+from dbus_fast.errors import DBusError
+from dbus_fast.message import Message
+from ops.charm import CharmBase
+from ops.framework import EventBase
+
+LIBID = "HPCTEAMUSEONLY"
+LIBAPI = 0
+LIBPATCH = 1
+PYDEPS = ["dbus-fast>=1.90.2"]
+
+_logger = logging.getLogger(__name__)
+_juju_unit = None
+_service_states = {}
+_service_hook_regex_filter = re.compile(r"service-(?P<service>[\w\\:-]*)-(?:started|stopped)")
+_DBUS_CHAR_MAPPINGS = MappingProxyType(
+    {
+        "_5f": "_",
+        "_40": "@",
+        "_2e": ".",
+        "_2d": "-",
+        "_5c": "\\",
+    }
+)
+
+
+def _systemctl(*args) -> None:
+    """Control systemd by via executed `systemctl ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if systemctl command fails.
+    """
+    cmd = ["systemctl", *args]
+    _logger.debug(f"systemd: Executing command {cmd}")
+    try:
+        subprocess.check_output(cmd)
+    except subprocess.CalledProcessError as e:
+        _logger.error(f"systemctl command failed: {e}")
+        raise
+
+
+_daemon_reload = functools.partial(_systemctl, "daemon-reload")
+_daemon_reload.__doc__ = "Reload systemd manager configuration."
+_start_service = functools.partial(_systemctl, "start")
+_start_service.__doc__ = "Start systemd service unit."
+_stop_service = functools.partial(_systemctl, "stop")
+_stop_service.__doc__ = "Stop systemd service unit."
+_enable_service = functools.partial(_systemctl, "enable")
+_enable_service.__doc__ = "Enable systemd service."
+_disable_service = functools.partial(_systemctl, "disable")
+_disable_service.__doc__ = "Disable systemd service."
```

Comment:
We already have a systemd library - how do folks feel about just introducing that as a dep here and using that?

Reply:
I would prefer that personally since then we're not on the hook for maintaining two separate implementations for interfacing with `systemctl` in `operator-libs-linux`. My only hang up however (and the reason why I initially implemented these methods) is that I am unsure how we would pull in the systemd charm library as a dependency. Can we use PYDEPS for this?

While we could _tell_ folks in the documentation for this charm lib that they must also consume the systemd charm library, I am concerned that if we make this workflow manual then someone will definitely bungle it by forgetting to pull in the systemd charm library before packing their charm.

Reply:
Yeh, that is a bit of a thing, but it's not the end of the world - in reality it would fail right away in unit tests etc.

The Observability team did this for a while with their `JujuTopology` lib - I think so long as it's properly documented in the docstring, it's probably the way to go :)

Reply:
I'd agree that would be a good trade-off if this was a complex chunk of code.

But here I'd much prefer the slight duplication to make it easier for users. As you note, charm libs can't have dependencies on each other as you (only on other PyPI libs via `PYDEPS`). As the [saying `go`es](https://go-proverbs.github.io/), "A little copying is better than a little dependency." It's easier for us to copy/maintain a few straight-forward lines of code (not much more than leftpad :-) than it will be for each consumer to remember to add the other charm lib.

Also, probably no need to copy the docstrings on these partials -- these are internal functions and pretty obvious. So that would save 5 lines of code. :-)